Added KSS features support including Config ID/SVN

hello31337 opened the pull request:
<!--
    Please fill in the following form before submitting this PR
    and ensure that your code follows our coding style guideline:
    https://gramine.readthedocs.io/en/latest/devel/coding-style.html -->

## Description of the changes <!-- (reasons and measures) -->
In the existing [PR#903](https://github.com/gramineproject/gramine/pull/903), an update was introduced to allow setting the KSS (Key Separation and Sharing) enable flag, as well as the ISV Family ID and ISV Extended Prod ID, in the manifest when KSS is enabled.

However, PR#903 did not provide functionality to set the Config ID and Config SVN, which are also available for use with KSS. Additionally, it appears that PR#903 has not yet been merged as of now.
Therefore, based on the updated codebase provided in PR#903, I have implemented a complete KSS support update that also allows specifying Config ID and Config SVN.

### Usage
The KSS enable flag, ISV Family ID, and ISV Extended Prod ID can be specified in the manifest in the same way as PR#903, as follows:

```
sgx.kss = true
sgx.isvfamilyid = "0x00112233445566778899aabbccddeeff"
sgx.isvextprodid = "0xcafef00dcafef00df00dcafef00dcafe"
```

Config ID and Config SVN are implemented by setting environment variables on the host OS, transferring them to the PAL, and registering the received values in SECS before ECREATE.  
Therefore, by setting the environment variables SGX_CONFIG_ID and SGX_CONFIG_SVN on the host OS in advance, you can specify the values you want to set.

```
export SGX_CONFIG_ID=DEADBEEF12345678DEADBEEF12345678DEADBEEF12345678DEADBEEF12345678DEADBEEF12345678DEADBEEF12345678DEADBEEF12345678DEADBEEF12345678
export SGX_CONFIG_SVN=ABCD
```

You can verify these values by running sgx-report.py or sgx-quote.py in CI-Examples/python/scripts:
```
$ gramine-sgx ./python scripts/sgx-quote.py 
Gramine is starting. Parsing TOML manifest file, this may take some time...
-----------------------------------------------------------------------------------------------------------------------
Gramine detected the following insecure configurations:

  - loader.insecure__use_cmdline_argv = true   (forwarding command-line args from untrusted host to the app)

Gramine will continue application execution, but this configuration must not be used in production!
-----------------------------------------------------------------------------------------------------------------------

Detected attestation type: dcap
Extracted SGX quote with size = 4734 and the following fields:
  ISVEXTPRODID:     cafef00dcafef00df00dcafef00dcafe
  ATTRIBUTES.FLAGS: 8500000000000000  [ Debug bit: False ]
  ATTRIBUTES.XFRM:  e700000000000000
  MRENCLAVE:        796e98b917f783f84886e77d2916eebb8e9a804e03d268035babfabf7c922c77
  MRSIGNER:         f4cd1ce4ed2384380df9a90020574994ad2d1db44d8d80a48a62f1c913273b1f
  CONFIGID:         deadbeef12345678deadbeef12345678deadbeef12345678deadbeef12345678
                    deadbeef12345678deadbeef12345678deadbeef12345678deadbeef12345678
  ISVPRODID:        0000
  ISVSVN:           0000
  CONFIGSVN:        cdab
  ISVFAMILYID:      00112233445566778899aabbccddeeff
  REPORTDATA:       0000000000000000000000000000000000000000000000000000000000000000
                    0000000000000000000000000000000000000000000000000000000000000000
```

### Known Issues
Currently, setting sgx.kss = false results in the following error:

```
gramine-sgx ./python scripts/sgx-quote.py 
Gramine is starting. Parsing TOML manifest file, this may take some time...
error: Enclave creation IOCTL failed with Input/output error (EIO). Probably your manifest requires CPU features (e.g. `sgx.cpu_features.avx512 = "required"`) that are not available on this platform.
error: Creating enclave failed: Input/output error (EIO)
error: load_enclave() failed with error: Input/output error (EIO)
```

Since I was prioritizing the implementation of KSS support and considering that it would be best to merge the update before significant changes occur in the existing codebase, I have submitted this PR without fixing the above issue.
I would greatly appreciate any assistance in resolving this issue.

<!--
    If your PR fixes an issue, please remember to add "Fixes #issue_number"
    here, to automatically close it on merge. -->

## How to test this PR? <!-- (if applicable) -->

mkow reviewed (changes requested):
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/2101)* status: 0 of 13 files reviewed, 12 unresolved discussions, not enough approvals from maintainers (3 more required), not enough approvals from different teams (2 more required, approved so far: ), period found at the end of a one-liner

___
*[a discussion](https://reviewable.io/reviews/gramineproject/gramine/2101#-OJi-IUs6S7b-rv-lnAx:-OJi-IUs6S7b-rv-lnAy:b-smgh1e) (no related file):*
> ### Known Issues
> 
> Currently, setting sgx.kss = false results in the following error:
> 
> ```
> gramine-sgx ./python scripts/sgx-quote.py 
> Gramine is starting. Parsing TOML manifest file, this may take some time...
> error: Enclave creation IOCTL failed with Input/output error (EIO). Probably your manifest requires CPU features (e.g. `sgx.cpu_features.avx512 = "required"`) that are not available on this platform.
> error: Creating enclave failed: Input/output error (EIO)
> error: load_enclave() failed with error: Input/output error (EIO)
> ```
> 
> Since I was prioritizing the implementation of KSS support and considering that it would be best to merge the update before significant changes occur in the existing codebase, I have submitted this PR without fixing the above issue.  
> I would greatly appreciate any assistance in resolving this issue.

Moving into a blocking comment, this has to be fixed before merging.
___
*[`-- commits` line 3 at r1](https://reviewable.io/reviews/gramineproject/gramine/2101#-OJhzzPRDOhtMJBBK76P:-OJhzzPRDOhtMJBBK76Q:undefined):*




_Suggestion:_
```
Add
```
___
*[`-- commits` line 5 at r1](https://reviewable.io/reviews/gramineproject/gramine/2101#-OJi-1UoCaQywQqhheku:-OJi-1UoCaQywQqhhekv:b7a792o):*
This should be squashed together with the first commit, it's an integral part of it.
___
*[`-- commits` line 7 at r1](https://reviewable.io/reviews/gramineproject/gramine/2101#-OJi-8QcCyPlKIxcD9aS:-OJi-8QcCyPlKIxcD9aT:undefined):*




_Suggestion:_
```
Fix
```
___
*[`CI-Examples/python/README.md` line 40 at r1](https://reviewable.io/reviews/gramineproject/gramine/2101#-OJi-XPC8doS0zFSDwmH:-OJi-XPC8doS0zFSDwmI:bkxn30) ([raw file](https://github.com/gramineproject/gramine/blob/6bbdd0d0eb36b99febf2314458dd9616b9a4ab33/CI-Examples/python/README.md#L40)):*
> ```Markdown
> ```
> 
> To set the Config ID and Config SVN, configure the following environment variables on the host OS before executing scripts:  
> ```

I don't like that those are passed through envs, why not via cmdline arguments?
___
*[`CI-Examples/python/README.md` line 40 at r1](https://reviewable.io/reviews/gramineproject/gramine/2101#-OJi-cfkBt0wOiKXBvUf:-OJi-cfkBt0wOiKXBvUg:b88ta8c) ([raw file](https://github.com/gramineproject/gramine/blob/6bbdd0d0eb36b99febf2314458dd9616b9a4ab33/CI-Examples/python/README.md#L40)):*
> ```Markdown
> ```
> 
> To set the Config ID and Config SVN, configure the following environment variables on the host OS before executing scripts:  
> ```

trailing spaces
___
*[`python/graminelibos/manifest.py` line 310 at r1](https://reviewable.io/reviews/gramineproject/gramine/2101#-OJi02O29YZ939DYzGsc:-OJi02O29YZ939DYzGsd:b671svd) ([raw file](https://github.com/gramineproject/gramine/blob/6bbdd0d0eb36b99febf2314458dd9616b9a4ab33/python/graminelibos/manifest.py#L310)):*
> ```Python
>         sgx.setdefault('kss', False)
>         sgx.setdefault('isvextprodid', "0x00000000000000000000000000000000")
>         sgx.setdefault('isvfamilyid', "0x00000000000000000000000000000000")
> ```

Are those two values interpreted as numbers or byte arrays? I think it's the latter? Then they shouldn't have the hex-number prefix "0x".
___
*[`pal/src/host/linux-sgx/host_main.c` line 204 at r1](https://reviewable.io/reviews/gramineproject/gramine/2101#-OJi-h9b8HKITeXzsvs-:-OJi-h9b8HKITeXzsvs0:b-k9o1hm) ([raw file](https://github.com/gramineproject/gramine/blob/6bbdd0d0eb36b99febf2314458dd9616b9a4ab33/pal/src/host/linux-sgx/host_main.c#L204)):*
> ```C
> 
> static int initialize_enclave(struct pal_enclave* enclave, const char* manifest_to_measure,
>                             uint8_t *config_id, uint16_t config_svn) {
> ```

Please fix formatting and the code style in the whole PR (see our style guideline in the docs)
___
*[`pal/src/host/linux-sgx/host_main.c` line 1135 at r1](https://reviewable.io/reviews/gramineproject/gramine/2101#-OJi1Cwi3w0ouaRrIn4t:-OJi1Cwj9sgwv3zTCzkL:b-p94wyg) ([raw file](https://github.com/gramineproject/gramine/blob/6bbdd0d0eb36b99febf2314458dd9616b9a4ab33/pal/src/host/linux-sgx/host_main.c#L1135)):*
> ```C
> }
> 
> // convert hexstring to byte array
> ```

Please don't write comment which say exactly what the next line says. You have basically the same sentence in the function name, so this comment doesn't help in anything.
___
*[`pal/src/host/linux-sgx/host_main.c` line 1136 at r1](https://reviewable.io/reviews/gramineproject/gramine/2101#-OJi19UTAXyzxqryTh62:-OJi19UTAXyzxqryTh63:b-svxbwa) ([raw file](https://github.com/gramineproject/gramine/blob/6bbdd0d0eb36b99febf2314458dd9616b9a4ab33/pal/src/host/linux-sgx/host_main.c#L1136)):*
> ```C
> 
> // convert hexstring to byte array
> static int parse_hex_string_to_bytes(const char* hex_str, uint8_t* buffer, size_t buffer_size) {
> ```

We already have `hex2bytes()`.
___
*[`pal/src/host/linux-sgx/host_main.c` line 1147 at r1](https://reviewable.io/reviews/gramineproject/gramine/2101#-OJi0hcL4UIMcgflG9Lr:-OJi0hcL4UIMcgflG9Ls:bp6hh1d) ([raw file](https://github.com/gramineproject/gramine/blob/6bbdd0d0eb36b99febf2314458dd9616b9a4ab33/pal/src/host/linux-sgx/host_main.c#L1147)):*
> ```C
>         buffer[i] = (uint8_t)strtol(byte_str, NULL, 16);
>     }
>     return 1;
> ```

Please use Gramine conventions, if you return status codes via an `int` then 0 means success and negative means error.
___
*[`pal/src/host/linux-sgx/host_main.c` line 1151 at r1](https://reviewable.io/reviews/gramineproject/gramine/2101#-OJi0cO01gsT7Del4_DS:-OJi0cO01gsT7Del4_DT:b-grb7ch) ([raw file](https://github.com/gramineproject/gramine/blob/6bbdd0d0eb36b99febf2314458dd9616b9a4ab33/pal/src/host/linux-sgx/host_main.c#L1151)):*
> ```C
> 
> // convert hexstring to uint16_t
> static int parse_hex_string_to_uint16(const char* hex_str, uint16_t* value) {
> ```

We already have `str_to_ulong()`, please use it.


<!-- Sent from Reviewable.io -->

kailun-qin reviewed (changes requested):
> Reviewed 13 of 13 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/2101)* status: all files reviewed, 24 unresolved discussions, not enough approvals from maintainers (2 more required), not enough approvals from different teams (2 more required, approved so far: ), period found at the end of a one-liner (waiting on @hello31337 and @mkow)

___
*[`CI-Examples/python/README.md` line 40 at r1](https://reviewable.io/reviews/gramineproject/gramine/2101#-OJi-XPC8doS0zFSDwmH:-OJqQx-mBD-YAzhlaxYq:b32v57) ([raw file](https://github.com/gramineproject/gramine/blob/6bbdd0d0eb36b99febf2314458dd9616b9a4ab33/CI-Examples/python/README.md#L40)):*
<details><summary><i>Previously, mkow (Michał Kowalczyk) wrote…</i></summary><blockquote>

I don't like that those are passed through envs, why not via cmdline arguments?
</blockquote></details>

+1
___
*[`python/graminelibos/manifest.py` line 310 at r1](https://reviewable.io/reviews/gramineproject/gramine/2101#-OJi02O29YZ939DYzGsc:-OJpwUBQ06A25uv9dTSG:b32v57) ([raw file](https://github.com/gramineproject/gramine/blob/6bbdd0d0eb36b99febf2314458dd9616b9a4ab33/python/graminelibos/manifest.py#L310)):*
<details><summary><i>Previously, mkow (Michał Kowalczyk) wrote…</i></summary><blockquote>

Are those two values interpreted as numbers or byte arrays? I think it's the latter? Then they shouldn't have the hex-number prefix "0x".
</blockquote></details>

+1
___
*[a discussion](https://reviewable.io/reviews/gramineproject/gramine/2101#-OJqDkso3rBvrVQPBmXy:-OJqDkso3rBvrVQPBmXz:b-kg5yr6) (no related file):*
I think we also need to update [`gramine-sgx-sigstruct-view`](https://github.com/gramineproject/gramine/blob/master/python/gramine-sgx-sigstruct-view) and the associated manpage: https://github.com/gramineproject/gramine/blob/master/Documentation/manpages/gramine-sgx-sigstruct-view.rst.
___
*[a discussion](https://reviewable.io/reviews/gramineproject/gramine/2101#-OJqH7vA6FhPqqAnq6Q3:-OJqH7vA6FhPqqAnq6Q4:b-fvewj8) (no related file):*
Shouldn't we update here: https://github.com/gramineproject/gramine/blob/ef48c72418b648dbbbb1d58a5e471eeab42d720d/pal/src/host/linux-sgx/pal_process.c#L108-L115?
___
*[a discussion](https://reviewable.io/reviews/gramineproject/gramine/2101#-OJqPITgD8CZgzjjkgX8:-OJqPITh68vIOLPvyo5G:b-nw1gza) (no related file):*
I think we also need to update our RA-TLS lib (e.g., the verification callback), the corresponding examples and the attestation doc.
___
*[a discussion](https://reviewable.io/reviews/gramineproject/gramine/2101#-OJqQ9Y1DjE2bMeDI098:-OJqQ9Y1DjE2bMeDI099:b50ljq1) (no related file):*
I think we should also note somewhere that setting `config_svn` impacts the sealing key derivation: https://github.com/gramineproject/gramine/blob/ef48c72418b648dbbbb1d58a5e471eeab42d720d/pal/src/host/linux-sgx/enclave_framework.c#L384-L385
___
*[`-- commits` line 4 at r1](https://reviewable.io/reviews/gramineproject/gramine/2101#-OJqDotp4-IqRpG2Ldr6:-OJqDotp4-IqRpG2Ldr7:b-96zehy):*
Pls add your sign-offs.
___
*[`python/graminelibos/manifest_check.py` line 108 at r1](https://reviewable.io/reviews/gramineproject/gramine/2101#-OJpwzWZ6PW2y0N9y9AE:-OJpwzWZ6PW2y0N9y9AF:bbt8k79) ([raw file](https://github.com/gramineproject/gramine/blob/6bbdd0d0eb36b99febf2314458dd9616b9a4ab33/python/graminelibos/manifest_check.py#L108)):*
> ```Python
>         'kss': bool,
>         'isvfamilyid': str,
>         'isvextprodid': str,
> ```

These should be placed above the TODO.



_Code quote:_
```Python
        'kss': bool,
        'isvfamilyid': str,
        'isvextprodid': str,
```
___
*[`pal/src/host/linux-sgx/host_main.c` line 1293 at r1](https://reviewable.io/reviews/gramineproject/gramine/2101#-OJqSLgt4ob6EYxC8UG4:-OJqSLguEKVc3oCqd90f:b-n1b8qo) ([raw file](https://github.com/gramineproject/gramine/blob/6bbdd0d0eb36b99febf2314458dd9616b9a4ab33/pal/src/host/linux-sgx/host_main.c#L1293)):*
> ```C
>         if (strcmp(key, "SGX_CONFIG_ID") == 0) {
>             if (!parse_hex_string_to_bytes(value, config_id, SGX_CONFIGID_SIZE)) {
>                 initialize_config_id(config_id);
> ```

I prefer that we fail loudly if parsing fails.



_Code quote:_
```C
initialize_config_id(config_id);
```
___
*[`pal/src/host/linux-sgx/host_main.c` line 1298 at r1](https://reviewable.io/reviews/gramineproject/gramine/2101#-OJqS_xY3-GQ0Gp0gSjj:-OJqS_xY3-GQ0Gp0gSjk:b-7ropln) ([raw file](https://github.com/gramineproject/gramine/blob/6bbdd0d0eb36b99febf2314458dd9616b9a4ab33/pal/src/host/linux-sgx/host_main.c#L1298)):*
> ```C
>         // parse SGX_CONFIG_SVN env to config_svn
>         else if (strcmp(key, "SGX_CONFIG_SVN") == 0) {
>             if (!parse_hex_string_to_uint16(value, &config_svn)) {
> ```

ditto
___
*[`CI-Examples/python/python.manifest.template` line 19 at r1](https://reviewable.io/reviews/gramineproject/gramine/2101#-OJpssaw6wcdZwNWLDzd:-OJpssaw6wcdZwNWLDze:bcgjvre) ([raw file](https://github.com/gramineproject/gramine/blob/6bbdd0d0eb36b99febf2314458dd9616b9a4ab33/CI-Examples/python/python.manifest.template#L19)):*
> ```
> loader.insecure__use_cmdline_argv = true
> 
> sys.enable_sigterm_injection = false
> ```

Why this change?



_Code quote:_
```
false
```
___
*[`CI-Examples/python/python.manifest.template` line 37 at r1](https://reviewable.io/reviews/gramineproject/gramine/2101#-OJpsuUN2yDBA7OQPHFR:-OJpsuUOBO0Xd9d1oItY:b-7ropln) ([raw file](https://github.com/gramineproject/gramine/blob/6bbdd0d0eb36b99febf2314458dd9616b9a4ab33/CI-Examples/python/python.manifest.template#L37)):*
> ```
> sys.enable_extra_runtime_domain_names_conf = true
> 
> sgx.debug = false
> ```

ditto



_Code quote:_
```
false
```
___
*[`CI-Examples/python/README.md` line 29 at r1](https://reviewable.io/reviews/gramineproject/gramine/2101#-OJpqMPA7mCCVdrV5GCC:-OJpqMPA7mCCVdrV5GCD:b88ta8c) ([raw file](https://github.com/gramineproject/gramine/blob/6bbdd0d0eb36b99febf2314458dd9616b9a4ab33/CI-Examples/python/README.md#L29)):*
> ```Markdown
> When KSS is enabled, you can set four values—ISV Family ID, ISV Extended Prod ID, Config ID, and Config SVN—and include them in the REPORT structure.
> 
> To enable KSS, modify the following section in python.manifest.template:  
> ```

trailing spaces



_Code quote:_
```Markdown
··
```
___
*[`CI-Examples/python/README.md` line 34 at r1](https://reviewable.io/reviews/gramineproject/gramine/2101#-OJpqPsjDyiWeOsOqfFy:-OJpqPsjDyiWeOsOqfFz:b-7ropln) ([raw file](https://github.com/gramineproject/gramine/blob/6bbdd0d0eb36b99febf2314458dd9616b9a4ab33/CI-Examples/python/README.md#L34)):*
> ```Markdown
> ```
> 
> To set the ISV Family ID and ISV Extended Prod ID, modify the following lines:  
> ```

ditto



_Code quote:_
```Markdown
··
```


<!-- Sent from Reviewable.io -->

hello31337 commented:
Thank you very much for your review. I am currently working on the modifications, and most of the changes you pointed out have already been addressed.

I would like to confirm one point regarding the modification to pass the Config ID and Config SVN as arguments.

Specifically, I believe that, due to the nature of the LibOS system calls such as fork and exec, it is difficult to pass these values to the subprocess via arguments.
If we were to pass them through arguments, it would likely require significant and incompatible changes to the LibOS system call interface and its implementation, which I don't believe would be welcomed by the Gramine community.

Therefore, if we pass the Config ID and Config SVN only to the main process via arguments and do not use environment variables, the following situation would occur:
* The main process would have the specified Config ID and Config SVN values.
* The subprocess, on the other hand, would have the default values (all zeros).

By the way, in Gramine, when spawning a subprocess, Local Attestation (LA) is performed between the main and subprocesses.
If we followe strict manners, during LA, the initiator of the REPORT needs to supply the target info of the responder as input to the EREPORT instruction.
However, until now, since the measurement (MRENCLAVE) of both initiator and responder were identical, meaning the target info was also identical, we simply used the local target info and didn't use peer's target info.
https://github.com/gramineproject/gramine/blob/v1.8/pal/src/host/linux-sgx/enclave_framework.c#L683-L688

This has worked fine so far, but with this PR's change, the Config ID and Config SVN will differ between the main and subprocess.
As a result, the LA REPORT verification fails, and the subprocess creation ends up failing.

At this point, I am considering three potential approaches for resolving this issue. Could you please advise which direction would be preferable from the Gramine project's perspective?

1. Pass the target info via IPC — Personally, I believe this is the recommended approach.
2. Temporarily zero out the [enclave info](https://github.com/gramineproject/gramine/blob/4629f068f68d2821607c6dc8d5d7502e0b2ab79a/pal/src/host/linux-sgx/pal_linux.h#L33), which acts as the source of the target info, during LA — I feel this is not a very elegant solution.
3. Modify the LibOS system call interface and implementation — Since this would involve breaking changes, I would prefer to avoid this if possible.

I would appreciate your guidance on this matter.

hello31337 commented:
Thank you very much for your review. Based on your comments, I have implemented a series of fixes and improvements.

Some parts remain unaddressed; this is a deliberate choice we discussed in a recent meeting, where we agreed to prioritize responsiveness and delegate certain minor issues for future handling on the Gramine side.

Below is a summary of the changes that have been applied, followed by items that are still pending. Please review at your convenience.

## Implemented Fixes and Improvements
* Fixed an issue where disabling KSS in the manifest led to ECREATE failure.

* Restored default values in the manifest to their original settings.

* Removed trailing whitespace from sentences in the README.

* Modified the code to accept Config ID and Config SVN via command-line arguments.

* Removed redundant comments.

* Replaced redundantly re-implemented logic with existing utility functions where applicable.

* Adjusted return value handling to follow Gramine coding conventions.

* Added explicit return statements in abnormal code paths where appropriate.

* Removed "0x" prefixes from initial values of ISV Extended Prod ID and ISV Family ID.

* Moved certain logic above TODO comments to improve code clarity.

* Enhanced gramine-kss-sigstruct-view to handle KSS values.

* Extended the RA-TLS verification callback mechanism to support KSS, and added a sample implementation: ra-tls-mbedtls-with-kss.

* Commit messages corrected and squashed.

Regarding the specification of Config ID and Config SVN: these values can now be provided either via command-line arguments or environment variables. Regardless of the method, the values are parsed in gramine.in and passed to host_main.c as command-line arguments. In other words, even when set via environment variables, they are internally parsed and forwarded as arguments.
If both command-line arguments and environment variables are specified simultaneously, a mismatch between the two will result in an error.
In the case where a subprocess is spawned, it inherits the Config ID and SVN values from the main process.

Regarding the callback mechanism: to support KSS without introducing breaking changes to existing products, a separate callback function was introduced with a modified interface.

## Remaining Items (Not Yet Addressed)
* Formatting has been partially adjusted, but not comprehensively reviewed.

* No explanation has been added regarding how the Config SVN value affects the sealing key. This might be more appropriate to document in a specific page on Gramine's ReadTheDocs.

* A detailed explanation of how to configure KSS in the manifest or at runtime is not yet provided; this too might be better placed in ReadTheDocs.

* The following change has not been applied: excluding Config ID/SVN from Report Data in both the is_peer_enclave_ok function and the Report Data generation logic: https://github.com/gramineproject/gramine/blob/v1.8/pal/src/host/linux-sgx/pal_process.c#L108-L110
